b2b/keycloak_admin_api.py

```diff
@@ -0,0 +1,427 @@
+
+
+class KeycloakAdminClient:
+    """Client for the Keycloak admin API."""
+
+    base_url = None
+    _realm = None
+    oauth_session = None
+    skip_verify = False
+
+    def __init__(self):
+        """
+        Configure the client so we can use it.
+
+        These settings must be in place:
+        - KEYCLOAK_BASE_URL: The base URL for the Keycloak instance.
+        - KEYCLOAK_DISCOVERY_URL: The OpenID discovery URL for the realm.
+        - KEYCLOAK_ADMIN_CLIENT_ID: The client ID to use for the admin client.
+        - KEYCLOAK_ADMIN_CLIENT_SECRET: The client secret to use for the admin client.
+        - KEYCLOAK_REALM_NAME: The realm we are working with.
+
+        If any of these are incorrect or missing, you'll get an
+        KeycloakAdminImproperlyConfiguredError.
+
+        The client should be one that has the ability to use the admin API. The
+        regular authentication client may not have this permission.
+        """
+
+        self.base_url = settings.KEYCLOAK_BASE_URL
+        if not self.base_url:
+            msg = "KEYCLOAK_BASE_URL setting is not configured."
+            raise KeycloakAdminImproperlyConfiguredError(msg)
+        self.base_url = urljoin(self.base_url, "/admin/realms/")
+        self._realm = settings.KEYCLOAK_REALM_NAME
+        if not self._realm:
+            msg = "KEYCLOAK_REALM_NAME setting is not configured."
+            raise KeycloakAdminImproperlyConfiguredError(msg)
+
+        realm_discovery = settings.KEYCLOAK_DISCOVERY_URL
+        if not realm_discovery:
+            msg = "KEYCLOAK_DISCOVERY_URL setting is not configured."
+            raise KeycloakAdminImproperlyConfiguredError(msg)
+        client_id = settings.KEYCLOAK_ADMIN_CLIENT_ID
+        if not client_id:
+            msg = "KEYCLOAK_ADMIN_CLIENT_ID setting is not configured."
+            raise KeycloakAdminImproperlyConfiguredError(msg)
+        client_secret = settings.KEYCLOAK_ADMIN_CLIENT_SECRET
+        if not client_secret:
+            msg = "KEYCLOAK_ADMIN_CLIENT_SECRET setting is not configured."
+            raise KeycloakAdminImproperlyConfiguredError(msg)
+
+        self.skip_verify = settings.KEYCLOAK_ADMIN_CLIENT_NO_VERIFY_SSL or False
+
+        openid_configuration = requests.get(
+            realm_discovery,
+            timeout=60,
+            verify=not self.skip_verify,
+        )
+        openid_configuration.raise_for_status()
+        self.openid_configuration = openid_configuration.json()
+
+        self.oauth_session = OAuth2Session(
+            client_id=client_id,
+            client_secret=client_secret,
+            token_endpoint=self.openid_configuration["token_endpoint"],
+            scope=settings.KEYCLOAK_ADMIN_CLIENT_SCOPES,
+            verify=not self.skip_verify,
+        )
+        self.token = self.oauth_session.fetch_token(
+            self.openid_configuration["token_endpoint"],
+            grant_type="client_credentials",
+        )
```

Comment:
I don't know if this could be potentially long-running, but if it might be it's probably worth hooking into the automatic refreshing mechanism: https://requests-oauthlib.readthedocs.io/en/latest/oauth2_workflow.html#third-recommended-define-automatic-token-refresh-and-update